docs/adding_abstract_interpretation_functions.md

```diff
@@ -0,0 +1,90 @@
+   don't get outdated if Torch changes an operator signature.
+
+3. Fill in the body of the function. Ideally this will just be a call
+   into a helper function from
+   [`torch/jit/_shape_functions.py`](https://github.com/pytorch/pytorch/blob/279634f384662b7c3a9f8bf7ccc3a6afd2f05657/torch/jit/_shape_functions.py#L1).
+   But in general, you will need to write the function and test it
+   (see the comments about "Shape, dtype, and decomposition function
+   testing infrastructure" in `testing_framework.py`). New shape
+   functions should be added upstream following the example of [this PR](https://github.com/pytorch/pytorch/pull/76889), 
+   though it can be useful to iterate locally in `abstract_interp_lib_gen.py` 
+   first.
+
+4. Re-run the `build_tools/update_abstract_interp_lib.sh` script to
+   update the library. After this step happens, ideally everything
+   "just works" and the functions are now correctly inferred for the
+   operator.
+
+## When things go wrong
+
+It is possible that the refinement pipeline (see [Shape and Dtype Refinement Pipeline Architecture](abstract_interp_lib.md#shape-and-dtype-refinement-pipeline-architecture))
+is not able to infer the shape/dtype of a tensor with a given
+abstract interpretation function. This usually means that there is something
+about the function which the optimizations in
+`torch-simplify-shape/dtype-functions`
+(`lib/Dialect/Torch/Transforms/Simplify{Shape/Dtype}Calculations.cpp`)
+cannot handle.
+
+To debug this, the overall goal is to pinpoint the IR construct that is not
+being simplified. This is usually accomplished by a combination of looking at
+the Python code for the function and the IR dumps. The best IR dump to look at
+varies, but frequently the IR dump right before `DropAbstractInterpCalculations`
+is the most useful, because it has already been simplified as much as possible,
+making it is easy to see what is blocking further simplification. Examples of
+issues you might see:
+
+- You might find that there is a loop with a non-constant trip count,
+  but based on your understanding of the function, you would expect it
+  to be simplified to a constant trip count -- you can then look at
+  the trip count calculation and see if there is a missing fold or
+  canonicalization.
+
+- You might find that there is a list operation that is not currently understood
+  by the optimizations. You can then teach the optimizations about that
+  operation.
+
+- You might find that there is an `Optional` value that you would
+  expect to be resolved to either a concrete value or `None`. You can
+  then look at the calculation that produces the optional value and
+  see what folds or canonicalizations are missing.
+
+See [this video](https://www.youtube.com/watch?v=E5epCJOtrf8) for general
+guidance on debugging Torch-MLIR.
+
+As a last resort, you can rewrite the function using constructs that
+`torch-simplify-shape/dtype-functions` can handle (look at other
```

Comment:
nit: I'm generally not a fan of these kinds of shortenings with `/` -- the reason is that then if you do a verbatim search for `torch-simplify-dtype-functions` you won't find this string. Even if longer, writing `torch-simplify-shape-functions/torch-simplify-dtype-functions` is superior.